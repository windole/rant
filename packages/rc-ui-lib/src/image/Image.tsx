import React, { CSSProperties, useRef, useMemo, useEffect, useContext } from 'react';
import classnames from 'classnames';
import { ImageProps } from './PropsType';
import { isDef, addUnit } from '../utils';
import Icon from '../icon';
import { useSetState, useUpdateEffect } from '../hooks';
import ConfigProviderContext from '../config-provider/ConfigProviderContext';

const Image: React.FC<ImageProps> = (props) => {
  const { prefixCls, createNamespace } = useContext(ConfigProviderContext);
  const [bem] = createNamespace('image', prefixCls);
  const [status, setStatus] = useSetState({ loading: true, error: false });
  const imgRef = useRef<HTMLImageElement>(null);
  const unmountedRef = useRef(false);

  const { fit, errorIcon, loadingIcon, showError, showLoading, block } = props;

  const style = useMemo(() => {
    const internalStyle: CSSProperties = {};

    if (isDef(props.width)) {
      internalStyle.width = addUnit(props.width);
    }

    if (isDef(props.height)) {
      internalStyle.height = addUnit(props.height);
    }

    if (isDef(props.radius)) {
      internalStyle.overflow = 'hidden';
      internalStyle.borderRadius = addUnit(props.radius);
    }

    return internalStyle;
  }, []);

  useEffect(() => {
    if (imgRef.current && imgRef.current.complete) {
      setStatus({ loading: false, error: false });
    }
    return () => {
      unmountedRef.current = true;
    };
  }, []);

  useUpdateEffect(() => {
    if (imgRef.current && !imgRef.current.complete) {
      setStatus({ error: false, loading: true });
    }
  }, [props.src]);

  const onLoad = (e) => {
    if (!unmountedRef.current) {
      setStatus({ loading: false });
      props.onLoad?.(e);
    }
  };

  const onError = (e) => {
    if (!unmountedRef.current) {
      setStatus({ loading: false, error: true });
      props.onError?.(e);
    }
  };

  const renderLoadingIcon = () => {
    if (typeof loadingIcon !== 'string') return loadingIcon;
    return (
      <Icon
        size={props.iconSize}
        name={loadingIcon}
        className={classnames(bem('loading-icon'))}
        classPrefix={props.iconPrefix}
      />
    );
  };

  const renderErrorIcon = () => {
    if (typeof errorIcon !== 'string') return errorIcon;
    return (
      <Icon
        size={props.iconSize}
        name={errorIcon}
        className={classnames(bem('error-icon'))}
        classPrefix={props.iconPrefix}
      />
    );
  };

  const renderPlaceholder = () => {
    if (status.loading && showLoading) {
      return (
        <div className={classnames(bem('loading'))} onClick={props.onClick}>
          {renderLoadingIcon()}
        </div>
      );
    }
    if (status.error && showError) {
      return (
        <div className={classnames(bem('error'))} onClick={props.onClick}>
          {renderErrorIcon()}
        </div>
      );
    }
    return null;
  };

  const renderImage = () => {
    if (status.error || !props.src) {
      return null;
    }

    const attrs = {
      className: classnames(bem('img')),
      style: {
        objectFit: fit,
      },
    };
    return (
      <img
        ref={imgRef}
        alt={props.alt || 'img'}
        src={props.src}
        onLoad={onLoad}
        onError={onError}
        {...attrs}
      />
    );
  };

  return (
    <div
      className={classnames(props.className, bem({ block, round: props.round }))}
      style={style}
      onClick={props.onClick}
    >
      {renderImage()}
      {renderPlaceholder()}
      {props.children}
    </div>
  );
};

Image.defaultProps = {
  fit: 'fill',
  errorIcon: 'photo-fail',
  loadingIcon: 'photo',
  showError: true,
  showLoading: true,
  block: true,
};

export default Image;
