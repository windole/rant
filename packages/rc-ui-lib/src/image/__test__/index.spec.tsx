import * as React from 'react';
import { mount } from 'enzyme';
import { fireEvent, render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import Image from '..';

const IMAGE_URL = 'https://img.com';

describe('Image', () => {
  afterEach(() => {
    jest.restoreAllMocks();
  });

  it('should emit load event after image loaded', async () => {
    const onLoad = jest.fn();
    const wrapper = mount(<Image src={IMAGE_URL} onLoad={onLoad} />);
    await wrapper.find('img').simulate('load');
    expect(onLoad).toHaveBeenCalled();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should watch src and reset', async () => {
    const wrapper = mount(<Image src={IMAGE_URL} />);
    await wrapper.find('img').simulate('load');
    await wrapper.setProps({ src: '' });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should emit error event when load image failed', () => {
    const onError = jest.fn();
    const { getByAltText } = render(<Image alt="errJpg" src={IMAGE_URL} onError={onError} />);
    fireEvent.error(getByAltText('errJpg'));
    expect(onError).toHaveBeenCalled();
  });

  it('should not render loading placeholder when show-loading prop is false', async () => {
    const wrapper = mount(<Image />);
    expect(wrapper.find('.rc-image__loading').exists()).toBeTruthy();
    await wrapper.setProps({
      showLoading: false,
    });
    expect(wrapper.find('.rc-image__loading').exists()).toBeFalsy();
  });

  it('should not render error placeholder when show-error prop is false', async () => {
    const { getByAltText, container } = render(
      <Image src={IMAGE_URL} alt="test" showError={false} />,
    );
    fireEvent.error(getByAltText('test'));
    expect(container.querySelector('.rc-image__error')).not.toBeInTheDocument();
  });

  it('should change loading icon when using loading-icon prop', () => {
    const wrapper = mount(<Image loadingIcon="success" />);
    expect(wrapper.find('.van-icon-success').exists()).toBeTruthy();
  });

  it('should render correctly when using width and height prop', () => {
    const { container } = render(<Image width={100} height={100} />);
    expect(container.querySelector('.rc-image').getAttribute('style')).toEqual(
      'width: 100px; height: 100px;',
    );
  });

  it('should render correctly when using radius prop', () => {
    const { container } = render(<Image width={100} height={100} radius={10} />);
    expect(container.querySelector('.rc-image')).toHaveStyle('border-radius: 10px');
  });
});
