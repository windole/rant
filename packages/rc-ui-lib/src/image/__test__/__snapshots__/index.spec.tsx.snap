// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image should emit load event after image loaded 1`] = `"<div class=\\"rc-image rc-image--block\\"><img alt=\\"img\\" src=\\"https://img.com\\" class=\\"rc-image__img\\" style=\\"object-fit: fill;\\"></div>"`;

exports[`Image should watch src and reset 1`] = `"<div class=\\"rc-image rc-image--block\\"></div>"`;
